test/reply-error.test.js

```diff
@@ -525,3 +525,73 @@ test('should trigger error handlers if a sync route throws any non-error object'
   t.equal(reply.statusCode, 500)
   t.equal(JSON.parse(reply.body).foo, 'bar')
 })
+
+// Issue 3283 https://github.com/fastify/fastify/issues/3283
+test('setting content-type on reply object should not hang the server case 1', t => {
+  t.plan(2)
+  const fastify = Fastify()
+
+  fastify.get('/', (req, reply) => {
+    reply
+      .code(200)
+      .headers({ 'content-type': 'text/plain; charset=utf-32' })
+      .send(JSON.stringify({ bar: 'foo', baz: 'foobar' }))
+  })
+
+  fastify.inject({
+    url: '/',
+    method: 'GET'
+  }, (err, res) => {
+    t.error(err)
+    t.equal(res.statusCode, 200)
+  })
+})
+
+test('setting content-type on reply object should not hang the server case 2', t => {
+  t.plan(2)
+  const fastify = Fastify()
+
+  fastify.get('/', (req, reply) => {
+    reply
+      .code(200)
+      .headers({ 'content-type': 'text/plain; charset=utf-8' })
+      .send({ bar: 'foo', baz: 'foobar' })
+  })
+
+  fastify.inject({
+    url: '/',
+    method: 'GET'
+  }, (err, res) => {
+    //  the server should not hang and send a 500 reply
+    t.error(err)
+    t.same(
+      {
+        error: 'Internal Server Error',
+        message: 'Attempted to send payload of invalid type \'object\'. Expected a string or Buffer.',
+        statusCode: 500,
+        code: 'FST_ERR_REP_INVALID_PAYLOAD_TYPE'
+      },
+      JSON.parse(res.payload)
+    )
+  })
+})
```

Comment:
I changed a bit this test to see more clearly that the server hangs out without your fix 👍🏽 

```suggestion
test('setting content-type on reply object should not hang the server case 2', async t => {
  t.plan(1)
  const fastify = Fastify()

  fastify.get('/', (req, reply) => {
    reply
      .code(200)
      .headers({ 'content-type': 'text/plain; charset=utf-8' })
      .send({ bar: 'foo', baz: 'foobar' })
  })

  try {
    await fastify.listen(0)
    const res = await fastify.inject({
      url: '/',
      method: 'GET'
    })
    t.same({
      error: 'Internal Server Error',
      message: 'Attempted to send payload of invalid type \'object\'. Expected a string or Buffer.',
      statusCode: 500,
      code: 'FST_ERR_REP_INVALID_PAYLOAD_TYPE'
    },
    res.json())
  } catch (error) {
    t.error(error)
  } finally {
    await fastify.close()
  }
})
```